### PR TITLE
[milvus-4.1.18] Enabling High Availability for Milvus 

### PR DESCRIPTION
**Pull request by Rachit-Chaudhary11**
## What this PR does / why we need it:
By enabling this feature, Milvus Instances will be resilient to failures, ensuring continuous availability and reliability for users. The implementation includes fault-tolerant mechanisms and redundancy configurations to mitigate downtime and main uninterrupted service in the event of node failures or network issues. Overall, this enhancement enhances the robustness and performance of Milvus, providing a more resilient infrastructure for data storage and retrieval.

## Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[mychartname]`)
- [x] PR only contains changes for one chart

### CODE REVIEW

**Comment by Rachit-Chaudhary11** [comment]
/assign @loveeachday
@haorenfsa please validate.

**Comment by haorenfsa** [comment]
context: #67

**Comment by haorenfsa** [comment]
Hi @Rachit-Chaudhary11, a little advice, better continue the work on the same PR, so that we could tract the context of the problem directly. 

For example, your previous work is on branch `rollingupgrade`, then you refine it on the branch `deploymentStrategy`.  
So after you finish your work, you can checkout `rollingupgrade` again, and `git reset deploymentStrategy --hard` to make this branch the same as `deploymentStrategy`. Then you can use `git push upstream rollingupgrade -f` the force update your `rollingupgrade` branch on github. 
In this way, we can now continue the progress on the previous PR opened with branch `rollingupgrade`.

**Comment by Rachit-Chaudhary11** [comment]
@haorenfsa 

> Hi @Rachit-Chaudhary11, a little advice, better continue the work on the same PR, so that we could tract the context of the problem directly.
> 
> For example, your previous work is on branch `rollingupgrade`, then you refine it on the branch `deploymentStrategy`. So after you finish your work, you can checkout `rollingupgrade` again, and `git reset deploymentStrategy --hard` to make this branch the same as `deploymentStrategy`. Then you can use `git push upstream rollingupgrade -f` the force update your `rollingupgrade` branch on github. In this way, we can now continue the progress on the previous PR opened with branch `rollingupgrade`.

I have done this to keep the alternative approaches in 2 separate branches as the other approach we are planning to use for some other feature. I hope you understand. Thanks

**Comment by Rachit-Chaudhary11** [comment]
> Hi @Rachit-Chaudhary11, a little advice, better continue the work on the same PR, so that we could tract the context of the problem directly.

@haorenfsa 
I have done this to keep the alternative approaches in 2 separate branches as the other approach we are planning to use for some other feature. I hope you understand. Thanks

**Comment by LoveEachDay** [comment]
@Rachit-Chaudhary11 Only if we have enabled activeStandby for milvus coordinator components, we can use `rollingUpdate` strategy, otherwise the pod will crash when rolling update.

Could you also check whether `rootCoordinator.avtiveStandby.enabled` is `true` before change the deployment strategy?

**Comment by Rachit-Chaudhary11** [comment]
@LoveEachDay kindly validate now.

**Comment by Rachit-Chaudhary11** [comment]
@LoveEachDay @haorenfsa kindly validate now.

**Comment by Rachit-Chaudhary11** [comment]
@LoveEachDay Thank you for the suggestion. I have the necessary changes, kindly validate. 
cc: @haorenfsa

File: charts/milvus/templates/mixcoord-deployment.yaml [hunk 1]
```diff
@@ -11,9 +11,18 @@ metadata:
 {{ include "milvus.ud.annotations" . | indent 4 }}
 
 spec:
+  {{- if ge (int .Values.mixCoordinator.replicas) 0 }}
```
Comment (haorenfsa) [inline charts/milvus/templates/mixcoord-deployment.yaml:14]: Sry, but I don't get it here. why is this `if` necessary? @Rachit-Chaudhary11

Comment (Rachit-Chaudhary11) [reply charts/milvus/templates/mixcoord-deployment.yaml:14]: It is available for all the components except the coordinators. Therefore I have added it.
For instance:
- https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/datanode-deployment.yaml#L15
- https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/indexnode-deployment.yaml#L14
- and so on

Comment (Rachit-Chaudhary11) [reply charts/milvus/templates/mixcoord-deployment.yaml:14]: @haorenfsa It is available for all the components except the coordinators. Therefore I have added it. 
For instance:
- https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/datanode-deployment.yaml#L15
- https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/indexnode-deployment.yaml#L14
- and so on

This is just a precautionary check.

Comment (Rachit-Chaudhary11) [reply charts/milvus/templates/mixcoord-deployment.yaml:14]: @haorenfsa It is available for all the components except the coordinators. Therefore I have added it. 
For instance:
- https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/datanode-deployment.yaml#L15
- https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/indexnode-deployment.yaml#L14
- and so on

This is just a precautionary check.

**Review by Rachit-Chaudhary11 — COMMENTED** [review]
This is a needed check to avoid any wrong replicas value to be inserted.

File: charts/milvus/templates/mixcoord-deployment.yaml (hunk 1 above)
Comment (Rachit-Chaudhary11) [reply charts/milvus/templates/mixcoord-deployment.yaml:14]: @haorenfsa It is available for all the components except the coordinators. Therefore I have added it.
For instance:

https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/datanode-deployment.yaml#L15
https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/indexnode-deployment.yaml#L14
and so on
This is just a precautionary check.

Comment (Rachit-Chaudhary11) [reply charts/milvus/templates/mixcoord-deployment.yaml:14]: @haorenfsa It is available for all the components except the coordinators. Therefore I have added it.
For instance:

https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/datanode-deployment.yaml#L15
https://github.com/zilliztech/milvus-helm/blob/master/charts/milvus/templates/indexnode-deployment.yaml#L14
and so on
This is just a precautionary check.

**Comment by Rachit-Chaudhary11** [comment]
@haorenfsa @LoveEachDay kindly validate the changes.

**Comment by Rachit-Chaudhary11** [comment]
@haorenfsa @LoveEachDay kindly let me know if there is anything else needed from my end.

File: charts/milvus/templates/mixcoord-deployment.yaml (hunk 1 above)
Comment (haorenfsa) [reply charts/milvus/templates/mixcoord-deployment.yaml:14]: ah, yes, it's for supporting hpa, #16

**Comment by haorenfsa** [comment]
@Rachit-Chaudhary11  Looks good. Thank you for the contribution!

**Pull request merged**